adding base_backup

sebasmannem opened the pull request:
Started working on adding BASE_BACKUP functionality.
More info, read https://www.postgresql.org/docs/current/protocol-replication.html

sebasmannem commented:
@jackc could you have a look at this PR?
Let me know if it needs anything, or raises questions.
Thanks :)

jackc commented:
I'm not very familiar with the base backup protocol so I can only really review it at a surface level. I'll add comments at the line level for them.

The long test is unfortunate, but I think its better than not being tested at all. However, it can be too long -- it appears to depend on the size of the database cluster.  I do most of my testing locally (no Docker or containers) so it tries to backup my local main database cluster that is quite large. It fails with a timeout. I suppose it really needs a fresh database cluster. It might be nice to note this in the README test setup instructions.

jackc commented on pglogrepl.go:
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
+	// Request a fast checkpoint.
+	Fast              bool
+	// Include the necessary WAL segments in the backup. This will include all the files between start and stop backup in the pg_wal directory of the base directory tar file.
+	Wal               bool
```
I think that this should be `WAL` to conform to Go naming conventions for acronyms.

sebasmannem replied:
Agree. CHanged.

jackc commented on pglogrepl.go:
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
+	// Specifying NOWAIT disables both the waiting and the warning, leaving the client responsible for ensuring the required log is available.
+	NoWait            bool
+	// Limit (throttle) the maximum amount of data transferred from server to client per unit of time (kb/s).
+	MaxRate           int
```
I think this is a known size field. I would think this should be `int32` or `int64`.

sebasmannem replied:
int32 is fine.

jackc commented on pglogrepl.go:
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
+
+// BaseBackupTablespace represents a tablespace in the backup
+type BaseBackupTablespace struct {
+	Oid int
```
`Oid` -> `OID`. OIDs are `uint32` in PostgreSQL.

sebasmannem replied:
Changed

jackc commented on pglogrepl.go:
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
+
+// BaseBackupResult will hold the return values  of the BaseBackup command
+type BaseBackupResult struct {
+	Lsn LSN
```
`Lsn` -> `LSN`

sebasmannem replied:
Changed

jackc commented on pglogrepl.go:
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
+	NoVerifyChecksums bool
+}
+
+func (bbo BaseBackupOptions) Sql() string {
```
`Sql` -> `SQL`

sebasmannem replied:
Yeah, but it could be private too, so I changed it to sql.

sebasmannem commented:
> The long test is unfortunate, but I think its better than not being tested at all....
I have added an option to disable the test if you want to.
> It fails with a timeout. 
I removed the timeouts.
> It might be nice to note this in the README test setup instructions.
I documented the option to disable it in the README file.

I also noticed some minor issues in tests (with tablespaces, and with higher incarnation), so I fixed them too.

jackc commented:
Looks pretty good. Just two more tweaks.

1. OID should be `uint32` instead of `int32`. Unlikely to cause any problems, but still better to be entirely correct.
2. There is a binary in one of the commits. If you can conveniently rebase / rewrite history to remove it and avoid bloating the repo that would be great. Otherwise, I can squash merge to remove it.

sebasmannem commented:
Ouch. Binary was in last commit. I removed it.
And changed int32 to uint32.

jackc commented:
Thanks!
